### ISSUE TITLE
Need help: playback stops when screen off?

### ISSUE THREAD

**Issue by ltGuillaume** [ltGuillaume/MusicFolderPlayer#19]
@sbncbot @OoLunar @Invictaz @xRadio-us
I found that for some people playback tends to stop on mobile devices when the screen is turned off. I'm not sure whether the cause is due to resource management of the device itself, or a crappy network connection. The playback tends to stop in between songs, never during, so it could be caused by JavaScript not being run when the screen is off.

- In between songs
- Screen is off (or the browser/just the music player's tab to the background?)
- Tends to immediately start playback again when I press the pause button :/

My question is whether you have experienced a similar issue and what the exact conditions were. Can you easily reproduce it? Also, if you can reproduce it, does enabling crossfade improve the situation?

Some links to information on a _possibly_ related issue:
https://stackoverflow.com/questions/22061280/javascript-halts-in-inactive-android-chrome-tab
https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API
https://stackoverflow.com/questions/15959244/is-it-possible-in-javascript-to-detect-when-the-screen-is-turned-off-in-the-an

**Comment by OoLunar** [comment]
On my app, it only turns off if it's on a different app, i.e the home page, or my contacts, etc. I have not tried enabling Crossfade, but I will take a look into it. The only solution I know is to turn off the screen while I'm still on my app.

**Comment by OoLunar** [comment]
Also, having a bad internet connection doesn't really seem to be an issue. I took some back roads for a trip, while choosing my songs with no reception, it kept playing (although I believe that's because it preloaded the song, and attempts to preload the next song while the current one playing. Very clever) almost all the way through

**Comment by OoLunar** [comment]
Service workers? 
https://stackoverflow.com/a/50309148

**Comment by OoLunar** [comment]
No sleep JavaScript? Looks more promising than it sounds.
https://github.com/richtr/NoSleep.js/blob/master/README.md

**Comment by ltGuillaume** [comment]
> On my app, it only turns off if it's on a different app, i.e the home page, or my contacts, etc. I have not tried enabling Crossfade, but I will take a look into it. The only solution I know is to turn off the screen while I'm still on my app.
> 
> EDIT: Enabling Crossfade does not help

Thanks for reporting! So, to be clear, you mean it will stop playback _after_ song X (not during song X?!) if you went to another app during playback of song X?

> Also, having a bad internet connection doesn't really seem to be an issue. I took some back roads for a trip, while choosing my songs with no reception, it kept playing (although I believe that's because it preloaded the song, and attempts to preload the next song while the current one playing. Very clever) almost all the way through

Yes, you can always use the interface after it's loaded, but it will load the songs only after the previous has finished playback. When crossfade is enabled, it will start playing the next song ~10 seconds before the current has finished. As such, there is no real preloading yet, but it is planned (by using the same code as for crossfade, but by staying paused until the previous song has finished).

> Service workers?
> https://stackoverflow.com/a/50309148

I have no experience with Service Workers yet (and in my desktop browser I have disabled them), but if the very first comment I read in your link is anything to go on, it's not really useful for this problem:

> SW are grate but not perfect, it's mostly limited to push notification, network event and caching things. It's _no good at staying alive_ and will force close after 10min or so if things are preventing it from going idle. It have _no control over playing music_, geo monitoring or p2p stuff :(

**Comment by OoLunar** [comment]
Shame SW don't work. Crossfade will just end the song normally while fading out.

**Comment by ltGuillaume** [comment]
> Shame SW don't work. Crossfade will just end the song normally while fading out.

That's so odd... First, if it really fades out, that means that JavaScript is still working. This could be because if audio is still being played, the browser will NOT stop JavaScript (`Tabs which are playing audio are considered foreground and aren’t throttled.` - [source](https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API)). But if that's the case, it should also load the next file AND start playback, all during playback of the last 10 seconds of the current song. If so, there will always be an audio track playing, so JavaScript can't be stopped. So why wouldn't it continue?

**Comment by ltGuillaume** [comment]
@OoLunar Could you tell me exactly how you provoke that it stops playback? I've been trying on an old Galaxy S5 while another app was opened, the screen off (and without crossfade), but it played for several hours without a problem (then I stopped it manually). I was using https://www.bromite.org as a browser on LineageOS 16 (Android 9).

Also, were you using the latest commit? (or just tell me the date of the music.* files, that should be enough of an indication for which commit you use).

**Comment by OoLunar** [comment]
May 4th.
As for the playback thing. Go and download [my app] (https://forsaken-borders.net/apps/), click a song of your choice, leave the app, and turn off the screen. By the end of the song, it stops.

**Comment by OoLunar** [comment]
Also, I know this is related what-so-ever to this issue, but can you help me with some Javascript related 10 line code? Send me an [email](mailto:lunar@forsaken-borders.net) if you want to help.

**Comment by ltGuillaume** [comment]
Thanks for the quick reply. Hmz, I'm guessing you need the actual Google Chrome for your app: when I click "listen to music" it just opens a browser pick dialog. When I pick Bromite (which is just an "ungoogled" Chrome), it still works (but not as a Custom Chrome Tab!): I started several other apps, then turned off the screen, and still it continued playback with at least 4 songs... It might depend on how aggressively the manufacturer has set the phone to kill background apps (more info at https://dontkillmyapp.com), but I just don't know for sure... What a mess...

**Comment by OoLunar** [comment]
That's rough then. I haven't have any clue on how to solve this.

**Comment by ltGuillaume** [comment]
I've installed Chrome on an old phone, let's see how it works. I did see that if I force-closed your app, playback also stopped. So perhaps you could try to exclude your app (and Chrome?) from battery optimization: https://www.makeuseof.com/tag/androids-battery-optimization-harm-good/

For now, it keeps playing on my phones.

**Comment by OoLunar** [comment]
Ah. I've already tried this on the Pixel 3. Doesn't seem to work. Instead I'm just am going to put the required .html, .js, and .php files into the app, basically turning it into a web app. Maybe then it'll work.

**Comment by ltGuillaume** [comment]
I don't know, will it? If you still use Custom Chrome Tabs, it basically still loads the browser. I've recently seen what you described on others' people phones: simply going from the browser into any other (memory hungry) app will close the entire browser and thus playback. Sucks bigtime... I might be able to reduce the memory footprint by "lazy-loading" the library tree, but I doubt that will change anything.

**Comment by ilPeo** [comment]
I confirm having the same problem on iOS 12.4 - tested on both iPhone and iPad, Safari and Firefox.
Everything works fine with any desktop browser on MacOS.
I love this project but this issue prevents me from using it everyday. Keep up the good work

**Comment by ltGuillaume** [comment]
I have no experience with iOS, but is there a way to allow "background data" for apps, and if so, is the browser allowed?

**Comment by ilPeo** [comment]
I don’t think so

**Comment by ilPeo** [comment]
A few searches on the Internet told me that's a Webkit/iOS problem. That sucks :(

**Comment by ltGuillaume** [comment]
It does indeed. I'm quite annoyed by this all, since it basically makes my efforts with MFP close to useless in so many use cases. I wish I'd have known this stuff beforehand. If you come by any possible workaround, even if it requires new code, please let me know.

Apart from that, I have not updated the code on GitHub with the new beta preload routine that I currently use. I'm not sure if this will do anything for this issue on iOS, but it could. If you wanna try it out, let me know.

**Comment by ilPeo** [comment]
> Il giorno 29 ago 2019, alle ore 20:35, Guillaume <notifications@github.com> ha scritto:
> 
> It does indeed. I'm quite annoyed by this all, since it basically makes my efforts with MFP close to useless in so many use cases. I wish I'd have known this stuff beforehand. If you come by any possible workaround, even if it requires new code, please let me know.
> 

I searched a little without finding any workaround, but I don’t know how to code, sorry.
The core of the issue seems to be how WebKit handles Serviceworkers.
> Apart from that, I have not updated the code on GitHub with the new beta preload routine that I currently use. I'm not sure if this will do anything for this issue on iOS, but it could. If you wanna try it out, let me know
> 
Worth trying, let me know how.

**Comment by ltGuillaume** [comment]
> but I don’t know how to code, sorry

I just meant I'm willing to spend time coding a workaround, if you'd find something that requires it. :-)

Here 3.0 beta: https://github.com/ltGuillaume/MusicFolderPlayer/tree/3.0

**Comment by ilPeo** [comment]
I’ll check it out in a few days and let you know.
Thanks

**Comment by ltGuillaume** [comment]
[FAQ](https://github.com/ltGuillaume/MusicFolderPlayer/wiki/FAQ)
Things to try, report back here afterwards.

**Comment by ltGuillaume** [comment]
@ilPeo Found anything? (also, see [FAQ](https://github.com/ltGuillaume/MusicFolderPlayer/wiki/FAQ))

**Comment by ilPeo** [comment]
I’ve been quite busy elsewhere and didn’t find anything useful to help you fix the problem

**Comment by Laynee** [comment]
It works fine with latest Firefox Android + Video Background Play Fix Addon.
Also works for YouTube playlists.

**Comment by ltGuillaume** [comment]
> 
> 
> It works fine with latest Firefox Android + Video Background Play Fix Addon. Also works for YouTube playlists.

Wow, thanks, I'll check it out. If it really does the trick, perhaps I can integrate whatever the add-on does into MFP.

**Comment by Laynee** [comment]
On older versions of Firefox Android, it even works without plugin.
The last working "old" version is [Firefox 68.11.0](https://archive.mozilla.org/pub/mobile/releases/68.11.0/). It's an ESR release from july 2020.

**Comment by ltGuillaume** [comment]
Yes, that corroborates my experience with Fennec (F-Droid): problem. Indeed, the switch to the new Fenic branch introduced some issues with continued playback.

**Comment by ltGuillaume** [comment]
@Laynee I just had a look at the code of https://github.com/mozilla/video-bg-play/blob/master/video-bg-play-content.js and this Video Background Play Fix add-on.

- First, it won't activate at all when you're using MusicFolderPlayer: it will only work for YouTube and Vimeo.
- Second, it seems to me that all this add-on does is prevent those sites themselves from stopping playback. It doesn't stop any browser behavior. So I don't think I can recreate any of this for MusicFolderPlayer.

**Comment by Laynee** [comment]
You're right the plugin doesn't help. Why I though it would I don't know. Sorry for the false hopes. But I've done a few tests on Android 10 and it's encouraging. The player works great with latest Firefox and Chrome.

The connexion was so poor the player couldn't even cache 10 seconds of a song before being disconnected. So it kept disconnecting, sometimes during 2 or 3 minutes but it always managed to reconnect, continue to play the current song, go to the next one etc. Screen off the whole time. That's impressive. Under the same conditions, YouTube usually crashes under 2 minutes. And I played 2 x 7 songs.

I'll try the player on other phones / browsers.

**Comment by ltGuillaume** [comment]
No worries, I'm just happy that you're helping out, plus I wanted to have a look-see at the inner workings of that add-on for a while now anyway.

As for my recent findings:

- Fennec on F-Droid (which is actually a Firefox _Fenix_ sans the telemetry), works OK, especially with stable connections, but can also stop playback while traveling. Some more retries in the code may be of help here.
- Ungoogled-Chromium just won't work: the next song will only start playback when the screen is on (or less than x minutes have passed, I think).
- Mull, a hardened Firefox (Fenix) fork, has serious problems loading the media, so it basically starts playback, doesn't finish the song, then skips 5 songs and stops playback. Really odd stuff.

**Comment by Shmik2003** [comment]
Maybe my message wouldn't satisfy the topic, but I met this problem and enabling Autoplay in settings of my browser helped.

I met this issue with another music server and thinked that it's problem of that server. MusicFolderPlayer gave an error that my browser blocks autoplay.

Now both are working and I think that I'll use MusicFolderPlayer.

**Comment by ltGuillaume** [comment]
@Shmik2003 Thanks for responding! Some additional questions:
- Which OS are you using?
- Which Browser are you using?
- Have  you tried how playback responds with the screen off?
